commcare_connect/opportunity/visit_import.py

```diff
@@ -106,16 +106,25 @@ def _bulk_update_visit_status(opportunity: Opportunity, dataset: Dataset):
 def update_payment_accrued(opportunity: Opportunity, users):
     payment_units = opportunity.paymentunit_set.all()
     for user in users:
-        user_visits = UserVisit.objects.filter(
-            opportunity=opportunity, user=user, status=VisitValidationStatus.approved
-        ).order_by("entity_id")
+        user_visits = (
+            UserVisit.objects.filter(opportunity=opportunity, user=user, status=VisitValidationStatus.approved)
+            .order_by("entity_id")
+            .values("entity_id", "deliver_unit_id")
+        )
         access = OpportunityAccess.objects.get(user=user, opportunity=opportunity)
         payment_accrued = 0
         for payment_unit in payment_units:
-            payment_unit_deliver_units = {deliver_unit.id for deliver_unit in payment_unit.deliver_units.all()}
-            for entity_id, visits in itertools.groupby(user_visits, key=lambda x: x.entity_id):
-                unit_counts = Counter(v.deliver_unit_id for v in visits)
-                number_completed = min(unit_counts[deliver_id] for deliver_id in payment_unit_deliver_units)
+            required_deliver_units = payment_unit.deliver_units.filter(optional=False).values_list("id", flat=True)
```

Comment:
Nit: this can be one db query and then filtered in python.

Reply:
Fixed. https://github.com/dimagi/commcare-connect/pull/282/commits/da1c32e22699d483ffb3db7c8b67f39d654d7fb0